pkg/i2gw/ingress2gateway.go

```diff
@@ -35,70 +35,135 @@ import (
 	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
 )
 
-func Run(printer printers.ResourcePrinter, namespace string, inputFile string) {
+// ProviderConstructorByName is a map of ProviderConstructor functions by a
+// provider name. Different Provider implementations should add their construction
+// func at startup.
+var ProviderConstructorByName = map[ProviderName]ProviderConstructor{}
+
+// ProviderName is a string alias that stores the concrete Provider name.
+type ProviderName string
+
+// ProviderConstructor is a construction function that constructs concrete
+// implementations of the Provider interface.
+type ProviderConstructor func(conf *ProviderConf) Provider
+
+// ProviderConf contains all the configuration required for every concrete
+// Provider implementation.
+type ProviderConf struct{}
+
+// The Provider interface specifies the required functionality which needs to be
+// implemented by every concrete Ingress/Gateway-API provider, in order for it to
+// be used.
+type Provider interface {
+	CustomResourceReader
+	ResourceConverter
+}
+
+type CustomResourceReader interface {
+
+	// ReadResourcesFromCluster reads custom resources associated with
+	// the underlying Provider implementation from the kubernetes cluster.
+	ReadResourcesFromCluster(ctx context.Context, customResources interface{}) error
+
+	// ReadResourcesFromFiles reads custom resources associated with
+	// the underlying Provider implementation from the files.
+	ReadResourcesFromFiles(ctx context.Context, customResources interface{}, filename string) error
+}
+
+// The ResourceConverter interface specifies all the implemented Gateway API resource
+// conversion functions.
+type ResourceConverter interface {
+
+	// ConvertHTTPRoutes converts the received ingresses and custom resources
+	// associated with the Provider into HTTPRoutes and Gateways.
+	ConvertHTTPRoutes(ingresses []networkingv1.Ingress, customResources interface{}) ([]gatewayv1beta1.HTTPRoute, []gatewayv1beta1.Gateway, field.ErrorList)
```

Comment:
should this be more generic (and maybe return a struct containing lists of xRoutes and Gateways) to support adding more Route types in the future (e.g. Contour can convert some HTTPProxies to TCP/TLSRoutes)

Reply:
I totally agree with you @sunjayBhatia, I thought of it too after the fact.
I will make the adjustment

Reply:
done